grammars/puppet.cson

```diff
@@ -394,23 +256,252 @@
-        'match': '(\\$)([a-zA-Zx7f-xff\\$]|::)([a-zA-Z0-9_x7f-xff\\$]|::)*\\b'
+        'match': '(\\${*)([a-zA-Zx7f-xff\\$]|::)([a-zA-Z0-9_x7f-xff\\$]|::)*\\b'
         'name': 'variable.other.readwrite.global.puppet'
       }
       {
+        'match': '(\\$\\{)(?:[a-zA-Zx7f-xff\\$]|::)(?:[a-zA-Z0-9_x7f-xff\\$]|::)*(\\})'
         'captures':
           '1':
             'name': 'punctuation.definition.variable.puppet'
           '2':
             'name': 'punctuation.definition.variable.puppet'
-        'match': '(\\$\\{)(?:[a-zA-Zx7f-xff\\$]|::)(?:[a-zA-Z0-9_x7f-xff\\$]|::)*(\\})'
         'name': 'variable.other.readwrite.global.puppet'
       }
     ]
+  'constants':
+    'begin': '((?:[-_A-Za-z0-9"\'.]+::)*[-_A-Za-z0-9"\'.]+)\\['
+    'beginCaptures':
+      '0':
+        'name': 'constant.support.puppet'
+    'end': '\\]'
+    'endCaptures':
+      '0':
+        'name': 'constant.support.puppet'
+    'patterns': [
+      {
+        'include': '#strings'
+      }
+      {
+        'include': '#variables'
+      }
+    ]
+  'conditionals':
+    'patterns': [
+      {
+        'include': '#comparisons'
+      }
+      {
+        'include': '#cases'
+      }
+    ]
+  'cases':
+    'begin': '(case)\\s+((\\${*)([a-zA-Zx7f-xff\\$]|::)([a-zA-Z0-9_x7f-xff\\$]|::)*\\b)\\s*({)'
+    'beginCaptures':
+      '1':
+        'name': 'keyword.control.puppet'
+      '2':
+        'name': 'variable.other.readwrite.global.puppet'
+      '3':
+        'name': 'punctuation.conditional.case.begin.puppet'
+    'end': '}'
+    'endCaptures':
+      '0':
+        'name': 'punctuation.conditional.case.end.puppet'
+    'name': 'meta.conditional.case.puppet'
+    'patterns': [
+      {
+        'include': '#case-parameters'
+      }
+    ]
+  'case-parameters':
+    'patterns': [
+      {
+        'begin': '(=>)\\s*({)'
+        'beginCaptures':
+          '0':
+            'name': 'punctuation.separator.key-value.puppet'
+          '2':
+            'name': 'punctuation.parameter.begin.puppet'
+        'end': '(})(?:,|$)'
+        'endCaptures':
+          '1':
+            'name': 'punctuation.parameter.end.puppet'
+        'patterns': [
+          {
+            'include': '#puppet'
+          }
+        ]
+      }
+      {
+        'include': '#variables'
+      }
+      {
+        'include': '#strings'
+      }
+    ]
+  'comparisons':
+    'begin': '(^|\\s)\\s*(if|else|unless|elsif)\\s+'
+    'beginCaptures':
+      '2':
+        'name': 'keyword.control.puppet'
+    'end': '{'
+    'patterns': [
+      {
+        'include': '#parameter-default-types'
+      }
+      {
+        'include': '#comparison-operators'
+      }
+      {
+        'include': '#negation-operators'
+      }
+    ]
+  'comparison-operators':
+    'match': '(?:\\s|^)(={2}|!=|>|<|>=|<=|=~|!~|in)(?:\\s|^)'
+    'name': 'keyword.control.puppet'
+  'negation-operators':
+    'captures':
+      '1':
+        'name': 'negation.keyword.control.puppet'
+    'match': '(?:\\s|^)(not|!)(?:\\s|^)'
```

Comment:
`match` above `captures` (same with the `numbers` match below this one), and same question regarding the ending `^` capture.